File: 216. Combination Sum III.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        results = []
        self.dfs(1, n, k, [], results)
        return results
        
        
    def dfs(self, start, dest, numleft, result, results):
        if dest == 0 and numleft == 0:
            results.append(result)
            return
        if dest != 0 and numleft <= 0:
            return
        for i in range(start,10):
            self.dfs(i+1, dest-i, numleft-1, result+[i], results)

File: 216. Combination Sum III.java
class Solution {
    List<List<Integer>> ans;
    public List<List<Integer>> combinationSum3(int k, int n) {
        ans = new ArrayList<>();
        dfs(1, k, 0, n, new ArrayList<>());
        return ans;
    }
    private void dfs(int start, int left, int sum, int n, List<Integer> ls) {
        if (left == 0) {
            if (sum == n) {
                ans.add(new ArrayList<>(ls));
            }
            return;
        }
        if (10-start < left) {
            return;
        } 
        if (sum + start > n) {
            return;
        }
        for (int i=start; i<10; i++) {
            ls.add(i);
            dfs(i+1, left-1, sum+i, n, ls);
            ls.remove(ls.size()-1);
        }
    }
}